faststreams/output_stream.nim

```diff
@@ -1,60 +1,94 @@
+    s*: OutputStream
 
-  OutputStreamVar* = ref OutputStream
+  AsyncOutputStream* {.borrow: `.`.} = distinct OutputStream
 
-  MemOutputStream* = distinct OutputStreamVar
-    # Writing to a MemOutputStream produces no side-effects
+  WritePageSyncProc* = proc (s: OutputStream, page: openarray[byte])
+                            {.nimcall, gcsafe, raises: [IOError, Defect].}
 
-  # Keep this temporary for backward-compatibility
-  DelayedWriteCursor* = WriteCursor
-  VarSizeWriteCursor* = distinct WriteCursor
+  WritePageAsyncProc* = proc (s: OutputStream, buf: pointer, bufLen: int): Future[void]
+                             {.nimcall, gcsafe, raises: [IOError, Defect].}
+
+  FlushSyncProc* = proc (s: OutputStream)
+                        {.nimcall, gcsafe, raises: [IOError, Defect].}
+
+  FlushAsyncProc* = proc (s: OutputStream): Future[void]
+                         {.nimcall, gcsafe, raises: [IOError, Defect].}
+
+  CloseSyncProc* = proc (s: OutputStream)
+                        {.nimcall, gcsafe, raises: [IOError, Defect].}
+
+  CloseAsyncProc* = proc (s: OutputStream): Future[void]
+                         {.nimcall, gcsafe, raises: [IOError, Defect].}
 
   OutputStreamVTable* = object
-    # TODO - the noSideEffect is temporary here until we switch to Nim 0.20.2
-    #        where noSideEffects overrides may make it possible to implement
-    #        the MemOutputStream handling.
-    writePage*: proc (s: OutputStreamVar, page: openarray[byte])
-                     {.noSideEffect, nimcall, gcsafe, raises: [IOError, Defect] .}
+    writePageSync*: WritePageSyncProc
+    writePageAsync*: WritePageAsyncProc
+    flushSync*: FlushSyncProc
+    flushAsync*: FlushAsyncProc
+    closeSync*: CloseSyncProc
+    closeAsyncProc*: CloseAsyncProc
 
-    flush*: proc (s: OutputStreamVar)
-                 {.noSideEffect, nimcall, gcsafe, raises: [IOError, Defect].}
+  WriteCursor* = object
+    head, bufferEnd: ptr byte
+    stream: OutputStream
 
-const
-  allocatorMetadata = 0 # TODO: Get this from Nim's allocator.
-                        # The goal is to make perfect page-aligned allocations
-  defaultPageSize = 4096 - allocatorMetadata - 1 # 1 byte for the null terminator
+  VarSizeWriteCursor* = distinct WriteCursor
 
-proc createWriteCursor*[R, T](x: var array[R, T]): WriteCursor =
-  let startAddr = cast[ptr byte](addr x[0])
-  WriteCursor(head: startAddr, bufferEnd: offset(startAddr, sizeof x))
+  FileOutputStream = ref object of OutputStream
+    file: File
 
-template canExtendOutput(s: OutputStreamVar): bool =
+const
+  nimAllocatorMetadataSize* = 0
+    # TODO: Get this from Nim's allocator.
+    # The goal is to make perfect page-aligned allocations
+  defaultPageSize = 4096 - nimAllocatorMetadataSize - 1 # 1 byte for the null terminator
+
+proc close*(s: var OutputStream) {.raises: [IOError, Defect].} =
+  if s != nil:
+    if s.vtable != nil and s.vtable.closeSync != nil:
+      s.vtable.closeSync(s)
+    s = nil
```

Comment:
Hmm, I have to think about this one. Why wouldn't you provide the same argument against the practice of using helpers such as `safe_delete` in C++?
https://stackoverflow.com/questions/540920/which-version-of-safe-delete-is-better

Reply:
well, one thing is the name: because `close` is a soft-close - `destroy` is what truly destroys it. in terms of balance, I imagine if I close something, I can reopen it - like a door (vs destroying the door). 

the other issue is more subtle: can we express meaningful values for all calls after `close`? if yes, not setting to `nil` makes it harder to fail using this API - by setting to nil, you're structurally introducing a hidden way to crash / fail.

the `safe_delete` case is a little bit the same - it's a relic from when `unique_ptr` didn't exist and you'd signal "deleted" values with null - this would allow you to manually check for deleted-or-not before using the pointer, but still not something to encourage - `safe_delete` would be the lesser evil because at least you "knew" the delete had happened. modern c++ will structurally encourage you not to expose null pointers and you won't explicitly delete either - fwiw, `unique_ptr` sets to null like `safe_delete` but the situation where you check for null almost never happens.

in C++ it's worse also - `if(this==null)` is UB so you can't write member functions that deal with `null` "gracefully". if we set to nil here, we should also ensure that _all_ proc's that take an outputstream check for nil before use (to remove the introduced hidden failure). If we use Nim `methods`, we can't however - these will crash on nil, and introducing `not nil` might be harder later if/when it starts working in the language.

Reply:
I think closing is a bit like nullifying the `unique_ptr` - you are not supposed to write the checks afterwards. The fact that you need a `var` value is a bit interesting as well, because it means that the compiler will let you call `close` easily only in the function where the stream was created (if you take it as an argument, you are working with a `let` value).

Using a `unique_ptr` after it was nullified is considered a Defect and the question is whether the same should be true for a stream. I think you have a reasonable point that `close` won't be seen in this way by most people, so if I can indeed make all operations safe after a close (which I think is possible), this might be better.